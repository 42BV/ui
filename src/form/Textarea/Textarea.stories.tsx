import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import { FieldTextarea, JarbTextarea, Textarea } from './Textarea';
import {
  FieldFormElementDependencies,
  FinalForm,
  JarbFormElementDependencies
} from '../../story-utils';
import { Alert } from 'reactstrap';
import { Card } from '../../card/Card/Card';
import { Tooltip } from '../../core/Tooltip/Tooltip';
import { Icon } from '../../core/Icon';

storiesOf('Form/Textarea', module)
  .addParameters({ component: Textarea })
  .addDecorator((Story) => (
    <>
      <Alert color="warning" className="mb-4">
        <p>
          To be able to use Textarea, you have to add react-textarea-autosize to
          your dependencies:
        </p>
        <code>npm install --save react-textarea-autosize</code>
      </Alert>
      <Story />
    </>
  ))
  .add('basic', () => {
    return (
      <Card className="m-2">
        <Textarea
          id="description"
          label="Description"
          placeholder="Please add a description"
          onChange={(value) => action(`onChange: ${value}`)}
        />
      </Card>
    );
  })
  .add('without placeholder', () => {
    return (
      <Card className="m-2">
        <Textarea
          id="description"
          label="Description"
          onChange={(value) => action(`onChange: ${value}`)}
        />
      </Card>
    );
  })
  .add('invisible label', () => {
    return (
      <Card className="m-2">
        <Textarea
          id="description"
          label="Description"
          hiddenLabel={true}
          placeholder="Please add a description"
          onChange={(value) => action(`onChange: ${value}`)}
        />
      </Card>
    );
  })
  .add('with custom label', () => {
    return (
      <Card className="m-2">
        <Textarea
          id="description"
          label={
            <div className="d-flex justify-content-between">
              <span>Description</span>
              <Tooltip
                className="ms-1"
                content="The description is shown inside a tooltip"
              >
                <Icon icon="info" />
              </Tooltip>
            </div>
          }
          placeholder="Please add a description"
          onChange={(value) => action(`onChange: ${value}`)}
        />
      </Card>
    );
  })
  .add('field', () => {
    return (
      <>
        <FieldFormElementDependencies />
        <FinalForm>
          <FieldTextarea
            id="description"
            name="description"
            label="Description"
            placeholder="Please add a description"
          />
        </FinalForm>
      </>
    );
  })
  .add('jarb', () => {
    return (
      <>
        <JarbFormElementDependencies />
        <FinalForm>
          <JarbTextarea
            id="description"
            name="description"
            placeholder="Please add a description"
            jarb={{
              validator: 'Hero.description',
              label: 'Description'
            }}
          />
        </FinalForm>
      </>
    );
  });
