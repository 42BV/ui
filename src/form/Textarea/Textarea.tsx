import React from 'react';
import { FormGroup, Label } from 'reactstrap';
import TextareaAutosize from 'react-textarea-autosize';
import classNames from 'classnames';

import withJarb from '../withJarb/withJarb';
import { Color } from '../types';

interface Props {
  /**
   * The id of the form element.
   */
  id: string;

  /**
   * The label of the form element.
   */
  label: string;

  /**
   * The placeholder of the form element.
   */
  placeholder?: string;

  /**
   * The value that the form element currently has.
   */
  value?: string;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: string) => void;

  /**
   * Optional callback for when the form element is blurred.
   */
  onBlur?: () => void;

  /**
   * Optional callback for when the form element is focused.
   */
  onFocus?: () => void;

  /**
   * Optionally the error message to render.
   */
  error?: React.ReactNode;

  /**
   * Optionally the color of the FormGroup.
   */
  color?: Color;

  /**
   * Whether or not the form element is currently valid.
   */
  valid?: boolean;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
}

/**
 * Textarea is a basic form element which allows the user to enter large
 * texts, autogrows when the user enters a lot of text.
 */
export default function Textarea(props: Props) {
  const {
    id,
    label,
    placeholder,
    onBlur,
    onFocus,
    onChange,
    value,
    color,
    error,
    valid,
    className = ''
  } = props;

  const inputProps = {
    id,
    value,
    placeholder,
    onChange: (event: { target: { value: string } }) =>
      onChange(event.target.value),
    onFocus,
    onBlur
  };

  const classes = classNames('form-control', { 'is-invalid': valid === false });

  return (
    <FormGroup className={className} color={color}>
      <Label for={id}>{label}</Label>
      <TextareaAutosize className={classes} {...inputProps} />
      {error}
    </FormGroup>
  );
}

/**
 * Variant of the Textarea which can be used in a Jarb context.
 */
export const JarbTextarea = withJarb<string, string, Props>(Textarea);
