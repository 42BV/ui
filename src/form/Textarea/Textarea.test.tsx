import React from 'react';
import { vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';

import { Textarea } from './Textarea';

describe('Component: Textarea', () => {
  function setup({
    value,
    hasPlaceholder,
    hasLabel
  }: {
    value?: string;
    hasPlaceholder?: boolean;
    hasLabel?: boolean;
  }) {
    const onChangeSpy = vi.fn();
    const onBlurSpy = vi.fn();
    const onFocusSpy = vi.fn();

    const props = {
      placeholder: hasPlaceholder ? 'Please enter your first name' : undefined,
      value,
      onChange: onChangeSpy,
      onBlur: onBlurSpy,
      onFocus: onFocusSpy,
      error: 'Some error',
      valid: true,
      id: hasLabel ? 'firstName' : undefined,
      label: 'First name',
      hiddenLabel: !hasLabel
    };

    const { container, rerender } = render(
      <Textarea color="success" {...props} />
    );

    return { container, props, rerender, onChangeSpy, onBlurSpy, onFocusSpy };
  }

  describe('ui', () => {
    test('default', () => {
      const { container } = setup({});
      expect(container).toMatchSnapshot();
    });

    test('with value', () => {
      setup({ value: 'Maarten' });
      // @ts-expect-error Form elements have property value
      expect(screen.getByRole('textbox').value).toEqual('Maarten');
    });

    test('with placeholder', async () => {
      expect.assertions(0);
      setup({ hasPlaceholder: true });
      await screen.findByPlaceholderText('Please enter your first name');
    });

    test('without placeholder', () => {
      setup({ hasPlaceholder: false });
      expect(
        screen.queryByPlaceholderText('Please enter your first name')
      ).toBeNull();
    });

    test('visible label', async () => {
      expect.assertions(0);
      setup({ hasLabel: true });
      await screen.findByText('First name');
      await screen.findByLabelText('First name');
    });

    test('invisible label', async () => {
      expect.assertions(1);
      setup({ hasLabel: false });
      expect(screen.queryByText('First name')).toBeNull();
      await screen.queryByLabelText('First name');
    });
  });

  describe('events', () => {
    test('onChange', () => {
      const { onChangeSpy } = setup({ value: undefined });

      fireEvent.change(screen.getByRole('textbox'), {
        target: { value: 'Maarten' }
      });

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith('Maarten');
    });

    test('onBlur', () => {
      const { onBlurSpy } = setup({ value: undefined });

      fireEvent.blur(screen.getByRole('textbox'));

      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });

    test('onFocus', () => {
      const { onFocusSpy } = setup({ value: undefined });

      fireEvent.focus(screen.getByRole('textbox'));

      expect(onFocusSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('value changes', () => {
    test('becomes empty', () => {
      const { props, rerender } = setup({ value: 'Maarten' });

      // @ts-expect-error Form elements have property value
      expect(screen.getByRole('textbox').value).toEqual('Maarten');

      const newProps = {
        ...props,
        value: ''
      };

      rerender(<Textarea {...newProps} />);

      // @ts-expect-error Form elements have property value
      expect(screen.getByRole('textbox').value).toEqual('');
    });

    test('becomes filled', () => {
      const { props, rerender } = setup({ value: undefined });

      // @ts-expect-error Form elements have property value
      expect(screen.getByRole('textbox').value).toEqual('');

      const newProps = {
        ...props,
        value: 'Maarten'
      };

      rerender(<Textarea {...newProps} />);

      // @ts-expect-error Form elements have property value
      expect(screen.getByRole('textbox').value).toEqual('Maarten');
    });
  });
});
