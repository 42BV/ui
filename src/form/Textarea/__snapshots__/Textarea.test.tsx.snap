// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Textarea ui default 1`] = `
<div>
  <div
    class="mb-3"
    color="success"
  >
    <textarea
      aria-label="First name"
      class="form-control"
      id="1"
    />
    Some error
  </div>
</div>
`;
